node/shard/src/lib.rs

```diff
@@ -174,25 +226,59 @@ impl ShardBlockChain {
         };
         let apply_result = self.runtime.write().apply(
             &apply_state,
-            &[],
+            &block.body.receipts,
             &block.body.transactions,
         );
         if apply_result.root != block.body.header.merkle_root_state {
-            info!(
+            error!(
                 "Merkle root {} is not equal to received {} after applying the transactions from {:?}",
-                block.body.header.merkle_root_state,
                 apply_result.root,
+                block.body.header.merkle_root_state,
                 block
             );
             false
+        } else if apply_result.new_receipts.is_empty() {
+            if let Some(merkle_root) = block.body.header.receipt_merkle_root {
+                error!(
+                    "Receipt Merkle root should not exist, but received {} after applying the transactions from {:?}",
+                    merkle_root,
+                    block
+                );
+                false
+            } else {
+                self.insert_block(
+                    &block,
+                    apply_result.db_changes,
+                    apply_result.tx_result,
+                    HashMap::new(),
+                );
+                true
+            }
         } else {
-            self.insert_block(
-                &block,
-                apply_result.db_changes,
-                apply_result.tx_result,
-                apply_result.new_receipts
-            );
-            true
+            let (shard_ids, new_receipts): (Vec<_>, Vec<_>) = apply_result.new_receipts
+                .into_iter()
+                .unzip();
+            let (receipt_merkle_root, receipt_merkle_paths) = merklize(&new_receipts);
```

Comment:
we duplicating bunch of code here, how about `apply_block` function just be:
 other_block = prepare_new_block(block.body.header.parent, block.body.transactions, ..);
 if other_block.body.header.state_merkle_root == block.body.header.state_merkle_root && other_block.body.header.receipt_merkle == block.body.heaer.receipt_merkle && .. {
    self.insert(other_block)
} else {
  error!
}